libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -213,19 +261,93 @@ void WaveTrack::Interval::StretchRightTo(double t)
+   const auto stretchRatio = GetStretchRatio();
+
+   auto success = false;
+   Finally Do { [&] {
+      if (!success)
+      {
+         TrimLeftTo(originalPlayStartTime);
+         TrimRightTo(originalPlayEndTime);
+      }
+   } };
+
+   // Leave 1 second of raw, unstretched audio before and after visible region
+   // to give the algorithm a chance to be in a steady state when reaching the
+   // play boundaries.
+   const auto tmpPlayStartTime =
+      std::max(GetSequenceStartTime(), originalPlayStartTime - stretchRatio);
+   const auto tmpPlayEndTime =
+      std::min(GetSequenceEndTime(), originalPlayEndTime + stretchRatio);
+   TrimLeftTo(tmpPlayStartTime);
+   TrimRightTo(tmpPlayEndTime);
+
+   WideClip wideClip { mpClip, mpClip1 };
+
+   constexpr auto sourceDurationToDiscard = 0.;
+   constexpr auto blockSize = 1024;
+   const auto numChannels = NChannels();
+   ClipTimeAndPitchSource stretcherSource { wideClip, sourceDurationToDiscard,
+                                            PlaybackDirection::forward };
+   TimeAndPitchInterface::Parameters params;
+   params.timeRatio = stretchRatio;
+   StaffPadTimeAndPitch stretcher { wideClip.GetRate(), numChannels, stretcherSource,
+                                    std::move(params) };
+
+   // Post-rendering sample counts, i.e., stretched units
+   const auto totalNumOutSamples =
+      sampleCount { wideClip.GetVisibleSampleCount().as_double() *
+                    stretchRatio };
+
+   sampleCount numOutSamples { 0 };
+   AudioContainer container(blockSize, numChannels);
+
+   const auto dst = std::make_shared<Interval>(
+      group, NChannels(), factory, GetClip(0)->GetRate());
+
+   while (numOutSamples < totalNumOutSamples)
+   {
+      const auto numSamplesToGet =
+         limitSampleBufferSize(blockSize, totalNumOutSamples - numOutSamples);
+      stretcher.GetSamples(container.Get(), numSamplesToGet);
+      constSamplePtr data[2];
+      data[0] = reinterpret_cast<constSamplePtr>(container.Get()[0]);
+      if (NChannels() == 2)
+         data[1] = reinterpret_cast<constSamplePtr>(container.Get()[1]);
+      // TODO I don't think so, but double-checking: the `excsafety{Strong}`
+      // comment of `Sequence::AppendNewBlock` doesn't mean the method can't
+      // throw, right? If it didn't throw, then we could get rid of the Finally
+      // and restore the trimming of `this` after this loop before the call to
+      // `Flush`.
+      dst->Append(data, numSamplesToGet);
+      numOutSamples += numSamplesToGet;
+      if (reportProgress)
+         reportProgress(
+            numOutSamples.as_double() / totalNumOutSamples.as_double());
+   }
+   dst->Flush();
+
+   // Now we're all like `this` except unstretched. We can clear leading and
+   // trailing, stretching transient parts.
+   dst->SetPlayStartTime(tmpPlayStartTime);
+   dst->ClearLeft(originalPlayStartTime);
+   dst->ClearRight(originalPlayEndTime);
+
+   // We don't preserve cutlines but the envelope.
```

Comment:
Cutlines?  Why not?

They might be preserved un-stretched.  Maybe nobody cares.

Reply:
This is a design change.  Make sure designers agree.

Expansion of cutlines is done by WaveClip::Paste in WaveClip::ExpandCutLine.  Paste fails if ratios don't match.

So I think the complete solution would be correct remapping of cutlines along the timeline at the time stretch is applied, but then computing the stretch of the cutline could be delayed until just before the Paste.

Reply:
> This is a design change. Make sure designers agree.

@dozzzzer logged a follow-up ticket that still awaits Martin's (or Leo's, but he's on vacation) input.